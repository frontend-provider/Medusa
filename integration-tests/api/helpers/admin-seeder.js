const Scrypt = require("scrypt-kdf")
const { User } = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const buf = await Scrypt.kdf("secret_password", { logN: 1, r: 1, p: 1 })
  const password_hash = buf.toString("base64")

  await manager.insert(User, {
    id: "admin_user",
    email: "admin@medusa.js",
    api_token: "test_token",
    role: "admin",
    password_hash,
    ...data,
  })
}
