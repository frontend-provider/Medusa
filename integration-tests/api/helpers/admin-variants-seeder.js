const {
  Region,
  Product,
  ProductVariant,
  PriceList,
  CustomerGroup,
  Customer,
  Image,
  ShippingProfile,
  ProductCollection,
  ProductOption,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const yesterday = ((today) => new Date(today.setDate(today.getDate() - 1)))(
    new Date()
  )

  const tenDaysAgo = ((today) => new Date(today.setDate(today.getDate() - 10)))(
    new Date()
  )

  const defaultProfile = await manager.findOne(ShippingProfile, {
    where: { type: ShippingProfile.default },
  })

  const collection = manager.create(ProductCollection, {
    id: "test-collection",
    handle: "test-collection",
    title: "Test collection",
  })

  await manager.save(collection)

  await manager.insert(Region, {
    id: "reg-europe",
    name: "Test Region Europe",
    currency_code: "eur",
    tax_rate: 0,
  })

  await manager.insert(Region, {
    id: "reg-us",
    name: "Test Region US",
    currency_code: "usd",
    tax_rate: 0,
  })

  const customer = await manager.create(Customer, {
    id: "test-customer",
    email: "john@doe.com",
    first_name: "John",
    last_name: "Doe",
    password_hash:
      "c2NyeXB0AAEAAAABAAAAAVMdaddoGjwU1TafDLLlBKnOTQga7P2dbrfgf3fB+rCD/cJOMuGzAvRdKutbYkVpuJWTU39P7OpuWNkUVoEETOVLMJafbI8qs8Qx/7jMQXkN", // password matching "test"
    has_account: true,
  })

  const customerGroup = await manager.create(CustomerGroup, {
    id: "test-group",
    name: "test-group",
  })

  await manager.save(customerGroup)
  customer.groups = [customerGroup]
  await manager.save(customer)

  const priceListActive = await manager.create(PriceList, {
    id: "pl",
    name: "VIP sale",
    description: "All year sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  await manager.save(priceListActive)

  const priceListExpired = await manager.create(PriceList, {
    id: "pl_expired",
    name: "VIP summer sale",
    description: "Summer sale for VIP customers.",
    type: "sale",
    status: "active",
    starts_at: tenDaysAgo,
    ends_at: yesterday,
  })

  await manager.save(priceListExpired)

  const priceListWithCustomers = await manager.create(PriceList, {
    id: "pl_with_customers",
    name: "VIP winter sale",
    description: "Winter sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  priceListWithCustomers.customer_groups = [customerGroup]

  await manager.save(priceListWithCustomers)

  const productMultiReg = manager.create(Product, {
    id: "test-product",
    handle: "test-product-reg",
    title: "Multi Reg Test product",
    profile_id: defaultProfile.id,
    description: "test-product-description",
    status: "published",
    collection_id: "test-collection",
  })

  const image = manager.create(Image, {
    id: "test-image",
    url: "test-image.png",
  })

  productMultiReg.images = [image]

  await manager.save(productMultiReg)

  await manager.save(ProductOption, {
    id: "test-option",
    title: "test-option",
    product_id: "test-product",
  })

  const variantMultiReg = await manager.create(ProductVariant, {
    id: "test-variant",
    inventory_quantity: 10,
    title: "Test variant",
    variant_rank: 0,
    sku: "test-sku",
    ean: "test-ean",
    upc: "test-upc",
    barcode: "test-barcode",
    product_id: "test-product",
    prices: [
      {
        id: "test-price-multi-usd",
        currency_code: "usd",
        type: "default",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-usd",
        currency_code: "usd",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-expired-multi-usd",
        currency_code: "usd",
        amount: 70,
        price_list_id: "pl_expired",
      },
      {
        id: "test-price-multi-eur",
        currency_code: "eur",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-eur",
        currency_code: "eur",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-multi-eur-with-customer",
        currency_code: "eur",
        amount: 40,
        price_list_id: "pl_with_customers",
      },
      {
        id: "test-price-discount-expired-multi-eur",
        currency_code: "eur",
        amount: 70,
        price_list_id: "pl_expired",
      },
    ],
    options: [
      {
        id: "test-variant-option-reg",
        value: "Default variant",
        option_id: "test-option",
      },
    ],
  })

  await manager.save(variantMultiReg)
}
