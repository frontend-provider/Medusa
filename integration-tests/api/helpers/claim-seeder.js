const {
  ClaimOrder,
  Order,
  LineItem,
  Fulfillment,
  Return,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  let orderWithClaim = manager.create(Order, {
    id: "order-with-claim",
    customer_id: "test-customer",
    email: "test@email.com",
    payment_status: "captured",
    fulfillment_status: "fulfilled",
    billing_address: {
      id: "test-billing-address",
      first_name: "lebron",
    },
    shipping_address: {
      id: "test-shipping-address",
      first_name: "lebron",
      country_code: "us",
    },
    region_id: "test-region",
    currency_code: "usd",
    tax_rate: 0,
    discounts: [],
    payments: [
      {
        id: "test-payment-for-claim-order",
        amount: 10000,
        currency_code: "usd",
        amount_refunded: 0,
        provider_id: "test-pay",
        data: {},
      },
    ],
    items: [],
    ...data,
  })

  await manager.save(orderWithClaim)

  const li = manager.create(LineItem, {
    id: "test-item-co-2",
    fulfilled_quantity: 1,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 1,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li)

  const li2 = manager.create(LineItem, {
    id: "test-item-co-3",
    fulfilled_quantity: 4,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 4,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li2)

  const claimWithFulfillment = manager.create(ClaimOrder, {
    id: "claim-w-f",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  const ful1 = manager.create(Fulfillment, {
    id: "fulfillment-co-1",
    data: {},
    provider_id: "test-ful",
  })

  const ful2 = manager.create(Fulfillment, {
    id: "fulfillment-co-2",
    data: {},
    provider_id: "test-ful",
  })

  claimWithFulfillment.fulfillments = [ful1, ful2]

  await manager.save(claimWithFulfillment)

  const claimWithReturn = manager.create(ClaimOrder, {
    id: "claim-w-r",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  await manager.save(claimWithReturn)

  await manager.insert(Return, {
    id: "return-id-2",
    claim_order_id: "claim-w-r",
    status: "requested",
    refund_amount: 0,
    data: {},
  })
}
