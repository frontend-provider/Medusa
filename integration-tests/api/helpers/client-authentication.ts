const AUTH_COOKIE = {}
export async function getClientAuthenticationCookie(
  api,
  email = null,
  password = null
) {
  const user = {
    email: email ?? "test@medusajs.com",
    password: password ?? "test",
  }

  if (AUTH_COOKIE[user.email]) {
    return AUTH_COOKIE[user.email]
  }

  const authResponse = await api.post("/store/auth", user)
  AUTH_COOKIE[user.email] = authResponse.headers["set-cookie"][0].split(";")

  return AUTH_COOKIE[user.email]
}
