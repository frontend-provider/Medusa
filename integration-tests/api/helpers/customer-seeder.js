const { Customer, Address } = require("@medusajs/medusa")

module.exports = async (connection, data = {}) => {
  const manager = connection.manager

  await manager.insert(Customer, {
    id: "test-customer-1",
    email: "test1@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-2",
    email: "test2@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-3",
    email: "test3@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-has_account",
    email: "test4@email.com",
    has_account: true,
  })

  await manager.insert(Address, {
    id: "test-address",
    first_name: "Lebron",
    last_name: "James",
    customer_id: "test-customer-1",
  })
}
