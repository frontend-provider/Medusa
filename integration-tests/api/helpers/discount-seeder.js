const {
  ShippingProfile,
  Region,
  Discount,
  DiscountRule,
} = require("@medusajs/medusa")
module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  await manager.insert(Region, {
    id: "test-region",
    name: "Test Region",
    currency_code: "usd",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })

  await manager.insert(Region, {
    id: "test-region-2",
    name: "Test Region 2",
    currency_code: "eur",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })
}
