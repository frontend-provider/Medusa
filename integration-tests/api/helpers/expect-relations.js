const { expect } = require("@jest/globals")

export function expectRelations(expected, actual){
    expected.forEach(a => expectRelations_(a, actual))
}

function expectRelations_(expected, actual){
    const items = expected.split(".");
    let data_ = actual;

    for(const i in items){
        data_ = data_[items[i]];
        
        if(data_ instanceof Array && data_.length > 0) {
            data_ = data_[0];
        } 
    
        expect(data_).toBeDefined();
    }
}
