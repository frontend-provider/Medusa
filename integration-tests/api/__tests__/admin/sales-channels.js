const path = require("path")

const { SalesChannel, Product } = require("@medusajs/medusa")

const { useApi } = require("../../../helpers/use-api")
const { useDb } = require("../../../helpers/use-db")

const adminSeeder = require("../../helpers/admin-seeder")
const {
  simpleSalesChannelFactory,
  simpleProductFactory,
  simpleCartFactory,
} = require("../../factories")
const { simpleOrderFactory } = require("../../factories")
const orderSeeder = require("../../helpers/order-seeder")
const productSeeder = require("../../helpers/product-seeder")

const startServerWithEnvironment =
  require("../../../helpers/start-server-with-environment").default

const adminReqConfig = {
  headers: {
    Authorization: "Bearer test_token",
  },
}

jest.setTimeout(50000)

describe("sales channels", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    const [process, connection] = await startServerWithEnvironment({
      cwd,
      env: { MEDUSA_FF_SALES_CHANNELS: true },
      verbose: false,
    })
    dbConnection = connection
    medusaProcess = process
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  describe("GET /admin/sales-channels/:id", () => {
    let salesChannel

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        salesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "test name",
          description: "test description",
        })
      } catch (e) {
        console.error(e)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("should retrieve the requested sales channel", async () => {
      const api = useApi()
      const response = await api.get(
        `/admin/sales-channels/${salesChannel.id}`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data.sales_channel).toBeTruthy()
      expect(response.data.sales_channel).toMatchSnapshot({
        id: expect.any(String),
        name: salesChannel.name,
        description: salesChannel.description,
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })
  })

  describe("GET /admin/sales-channels", () => {
    let salesChannel1
    let salesChannel2

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        salesChannel1 = await simpleSalesChannelFactory(dbConnection, {
          name: "test name",
          description: "test description",
        })
        salesChannel2 = await simpleSalesChannelFactory(dbConnection, {
          name: "test name 2",
          description: "test description 2",
        })
      } catch (e) {
        console.error(e)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("should list the sales channel", async () => {
      const api = useApi()
      const response = await api.get(`/admin/sales-channels/`, adminReqConfig)

      expect(response.status).toEqual(200)
      expect(response.data.sales_channels).toBeTruthy()
      expect(response.data.sales_channels.length).toBe(2)
      expect(response.data).toMatchSnapshot({
        count: 2,
        limit: 20,
        offset: 0,
        sales_channels: expect.arrayContaining([
          {
            id: expect.any(String),
            name: salesChannel1.name,
            description: salesChannel1.description,
            is_disabled: false,
            deleted_at: null,
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
          {
            id: expect.any(String),
            name: salesChannel2.name,
            description: salesChannel2.description,
            is_disabled: false,
            deleted_at: null,
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
        ]),
      })
    })

    it("should list the sales channel using free text search", async () => {
      const api = useApi()
      const response = await api.get(
        `/admin/sales-channels/?q=2`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data.sales_channels).toBeTruthy()
      expect(response.data.sales_channels.length).toBe(1)
      expect(response.data).toMatchSnapshot({
        count: 1,
        limit: 20,
        offset: 0,
        sales_channels: expect.arrayContaining([
          {
            id: expect.any(String),
            name: salesChannel2.name,
            description: salesChannel2.description,
            is_disabled: false,
            deleted_at: null,
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
        ]),
      })
    })

    it("should list the sales channel using properties filters", async () => {
      const api = useApi()
      const response = await api.get(
        `/admin/sales-channels/?name=test+name`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data.sales_channels).toBeTruthy()
      expect(response.data.sales_channels.length).toBe(1)
      expect(response.data).toMatchSnapshot({
        count: 1,
        limit: 20,
        offset: 0,
        sales_channels: expect.arrayContaining([
          {
            id: expect.any(String),
            name: salesChannel1.name,
            description: salesChannel1.description,
            is_disabled: false,
            deleted_at: null,
            created_at: expect.any(String),
            updated_at: expect.any(String),
          },
        ]),
      })
    })
  })

  describe("POST /admin/sales-channels/:id", () => {
    let sc

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        sc = await simpleSalesChannelFactory(dbConnection, {
          name: "test name",
          description: "test description",
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("updates sales channel properties", async () => {
      const api = useApi()

      const payload = {
        name: "updated name",
        description: "updated description",
        is_disabled: true,
      }

      const response = await api.post(
        `/admin/sales-channels/${sc.id}`,
        payload,
        {
          headers: {
            authorization: "Bearer test_token",
          },
        }
      )

      expect(response.status).toEqual(200)
      expect(response.data.sales_channel).toMatchSnapshot({
        id: expect.any(String),
        name: payload.name,
        description: payload.description,
        is_disabled: payload.is_disabled,
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })
  })

  describe("POST /admin/sales-channels", () => {
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
      } catch (e) {
        console.error(e)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("successfully creates a disabled sales channel", async () => {
      const api = useApi()

      const newSalesChannel = {
        name: "sales channel name",
        is_disabled: true,
      }

      const response = await api
        .post("/admin/sales-channels", newSalesChannel, adminReqConfig)
        .catch((err) => {
          console.log(err)
        })

      expect(response.status).toEqual(200)
      expect(response.data.sales_channel).toBeTruthy()

      expect(response.data).toMatchSnapshot({
        sales_channel: expect.objectContaining({
          name: newSalesChannel.name,
          is_disabled: true,
        }),
      })
    })

    it("successfully creates a sales channel", async () => {
      const api = useApi()

      const newSalesChannel = {
        name: "sales channel name",
        description: "sales channel description",
      }

      const response = await api
        .post("/admin/sales-channels", newSalesChannel, adminReqConfig)
        .catch((err) => {
          console.log(err)
        })

      expect(response.status).toEqual(200)
      expect(response.data.sales_channel).toBeTruthy()

      expect(response.data).toMatchSnapshot({
        sales_channel: expect.objectContaining({
          name: newSalesChannel.name,
          description: newSalesChannel.description,
          is_disabled: false,
        }),
      })
    })
  })

  describe("DELETE /admin/sales-channels/:id", () => {
    let salesChannel

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        salesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "test name",
          description: "test description",
        })

        await simpleSalesChannelFactory(dbConnection, {
          name: "Default channel",
          id: "test-channel",
          is_default: true,
        })
      } catch (e) {
        console.error(e)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("should delete the requested sales channel", async () => {
      const api = useApi()

      let deletedSalesChannel = await dbConnection.manager.findOne(
        SalesChannel,
        {
          where: { id: salesChannel.id },
          withDeleted: true,
        }
      )

      expect(deletedSalesChannel.id).toEqual(salesChannel.id)
      expect(deletedSalesChannel.deleted_at).toEqual(null)

      const response = await api.delete(
        `/admin/sales-channels/${salesChannel.id}`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data).toMatchSnapshot({
        deleted: true,
        id: expect.any(String),
        object: "sales-channel",
      })

      deletedSalesChannel = await dbConnection.manager.findOne(SalesChannel, {
        where: { id: salesChannel.id },
        withDeleted: true,
      })

      expect(deletedSalesChannel.id).toEqual(salesChannel.id)
      expect(deletedSalesChannel.deleted_at).not.toEqual(null)
    })

    it("should delete the requested sales channel idempotently", async () => {
      const api = useApi()

      let deletedSalesChannel = await dbConnection.manager.findOne(
        SalesChannel,
        {
          where: { id: salesChannel.id },
          withDeleted: true,
        }
      )

      expect(deletedSalesChannel.id).toEqual(salesChannel.id)
      expect(deletedSalesChannel.deleted_at).toEqual(null)

      let response = await api.delete(
        `/admin/sales-channels/${salesChannel.id}`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data).toEqual({
        id: expect.any(String),
        object: "sales-channel",
        deleted: true,
      })

      deletedSalesChannel = await dbConnection.manager.findOne(SalesChannel, {
        where: { id: salesChannel.id },
        withDeleted: true,
      })

      expect(deletedSalesChannel.id).toEqual(salesChannel.id)
      expect(deletedSalesChannel.deleted_at).not.toEqual(null)

      response = await api.delete(
        `/admin/sales-channels/${salesChannel.id}`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data).toEqual({
        id: expect.any(String),
        object: "sales-channel",
        deleted: true,
      })

      deletedSalesChannel = await dbConnection.manager.findOne(SalesChannel, {
        where: { id: salesChannel.id },
        withDeleted: true,
      })

      expect(deletedSalesChannel.id).toEqual(salesChannel.id)
      expect(deletedSalesChannel.deleted_at).not.toEqual(null)
    })

    it("should throw if we attempt to delete default channel", async () => {
      const api = useApi()
      expect.assertions(2)

      const res = await api
        .delete(`/admin/sales-channels/test-channel`, adminReqConfig)
        .catch((err) => err)

      expect(res.response.status).toEqual(400)
      expect(res.response.data.message).toEqual(
        "You cannot delete the default sales channel"
      )
    })
  })

  describe("GET /admin/orders/:id", () => {
    let order
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)

        order = await simpleOrderFactory(dbConnection, {
          sales_channel: {
            name: "test name",
            description: "test description",
          },
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("expands sales channel for single", async () => {
      const api = useApi()

      const response = await api.get(
        `/admin/orders/${order.id}`,
        adminReqConfig
      )

      expect(response.data.order.sales_channel).toBeTruthy()
      expect(response.data.order.sales_channel).toMatchSnapshot({
        id: expect.any(String),
        name: "test name",
        description: "test description",
        is_disabled: false,
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })
  })

  describe("GET /admin/orders?expand=sales_channels", () => {
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)

        await simpleOrderFactory(dbConnection, {
          sales_channel: {
            name: "test name",
            description: "test description",
          },
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("expands sales channel with parameter", async () => {
      const api = useApi()

      const response = await api.get(
        "/admin/orders?expand=sales_channel",
        adminReqConfig
      )

      expect(response.data.orders[0].sales_channel).toBeTruthy()
      expect(response.data.orders[0].sales_channel).toMatchSnapshot({
        id: expect.any(String),
        name: "test name",
        description: "test description",
        is_disabled: false,
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })
  })

  describe("GET /admin/product/:id", () => {
    let product
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)

        product = await simpleProductFactory(dbConnection, {
          sales_channels: [
            {
              name: "webshop",
              description: "Webshop sales channel",
            },
            {
              name: "amazon",
              description: "Amazon sales channel",
            },
          ],
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("returns product with sales channel", async () => {
      const api = useApi()

      const response = await api
        .get(`/admin/products/${product.id}`, adminReqConfig)
        .catch((err) => console.log(err))

      expect(response.data.product.sales_channels).toBeTruthy()
      expect(response.data.product.sales_channels).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            name: "webshop",
            description: "Webshop sales channel",
            is_disabled: false,
          }),
          expect.objectContaining({
            name: "amazon",
            description: "Amazon sales channel",
            is_disabled: false,
          }),
        ])
      )
    })
  })

  describe("GET /admin/products?expand[]=sales_channels", () => {
    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)

        await simpleProductFactory(dbConnection, {
          sales_channels: [
            {
              name: "webshop",
              description: "Webshop sales channel",
            },
            {
              name: "amazon",
              description: "Amazon sales channel",
            },
          ],
        })
      } catch (err) {
        console.log(err)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("expands sales channel with parameter", async () => {
      const api = useApi()

      const response = await api.get(
        "/admin/products?expand=sales_channels",
        adminReqConfig
      )

      expect(response.data.products[0].sales_channels).toBeTruthy()
      expect(response.data.products[0].sales_channels).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            name: "webshop",
            description: "Webshop sales channel",
            is_disabled: false,
          }),
          expect.objectContaining({
            name: "amazon",
            description: "Amazon sales channel",
            is_disabled: false,
          }),
        ])
      )
    })
  })

  describe("DELETE /admin/sales-channels/:id/products/batch", () => {
    let salesChannel
    let product

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        product = await simpleProductFactory(dbConnection, {
          id: "product_1",
          title: "test title",
        })
        salesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "test name",
          description: "test description",
          products: [product],
        })
      } catch (e) {
        console.error(e)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("should remove products from a sales channel", async () => {
      const api = useApi()

      let attachedProduct = await dbConnection.manager.findOne(Product, {
        where: { id: product.id },
        relations: ["sales_channels"],
      })

      expect(attachedProduct.sales_channels.length).toBe(1)
      expect(attachedProduct.sales_channels).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: expect.any(String),
            name: "test name",
            description: "test description",
            is_disabled: false,
          }),
        ])
      )

      const payload = {
        product_ids: [{ id: product.id }],
      }

      await api.delete(
        `/admin/sales-channels/${salesChannel.id}/products/batch`,
        {
          ...adminReqConfig,
          data: payload,
        }
      )
      // Validate idempotency
      const response = await api.delete(
        `/admin/sales-channels/${salesChannel.id}/products/batch`,
        {
          ...adminReqConfig,
          data: payload,
        }
      )

      expect(response.status).toEqual(200)
      expect(response.data.sales_channel).toEqual(
        expect.objectContaining({
          id: expect.any(String),
          name: "test name",
          description: "test description",
          is_disabled: false,
        })
      )

      attachedProduct = await dbConnection.manager.findOne(Product, {
        where: { id: product.id },
        relations: ["sales_channels"],
      })

      expect(attachedProduct.sales_channels.length).toBe(0)
    })
  })

  describe("POST /admin/sales-channels/:id/products/batch", () => {
    let salesChannel
    let product

    beforeEach(async () => {
      try {
        await adminSeeder(dbConnection)
        salesChannel = await simpleSalesChannelFactory(dbConnection, {
          name: "test name",
          description: "test description",
        })
        product = await simpleProductFactory(dbConnection, {
          id: "product_1",
          title: "test title",
        })
      } catch (e) {
        console.error(e)
      }
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("should add products to a sales channel", async () => {
      const api = useApi()

      const payload = {
        product_ids: [{ id: product.id }],
      }

      const response = await api.post(
        `/admin/sales-channels/${salesChannel.id}/products/batch`,
        payload,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data.sales_channel).toEqual({
        id: expect.any(String),
        name: "test name",
        description: "test description",
        is_disabled: false,
        created_at: expect.any(String),
        updated_at: expect.any(String),
        deleted_at: null,
      })

      const attachedProduct = await dbConnection.manager.findOne(Product, {
        where: { id: product.id },
        relations: ["sales_channels"],
      })

      expect(attachedProduct.sales_channels.length).toBe(1)
      expect(attachedProduct.sales_channels).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: expect.any(String),
            name: "test name",
            description: "test description",
            is_disabled: false,
          }),
        ])
      )
    })
  })

  describe("/admin/orders using sales channels", () => {
    describe("GET /admin/orders", () => {
      let order

      beforeEach(async () => {
        try {
          await adminSeeder(dbConnection)
          order = await simpleOrderFactory(dbConnection, {
            sales_channel: {
              name: "test name",
              description: "test description",
            },
          })
          await orderSeeder(dbConnection)
        } catch (err) {
          console.log(err)
          throw err
        }
      })

      afterEach(async () => {
        const db = useDb()
        await db.teardown()
      })

      it("should successfully lists orders that belongs to the requested sales channels", async () => {
        const api = useApi()

        const response = await api.get(
          `/admin/orders?sales_channel_id[]=${order.sales_channel_id}`,
          {
            headers: {
              authorization: "Bearer test_token",
            },
          }
        )

        expect(response.status).toEqual(200)
        expect(response.data.orders.length).toEqual(1)
        expect(response.data.orders).toEqual([
          expect.objectContaining({
            id: order.id,
          }),
        ])
      })
    })
  })

  describe("/admin/products using sales channels", () => {
    describe("GET /admin/products", () => {
      const productData = {
        id: "product-sales-channel-1",
        title: "test description",
      }
      let salesChannel

      beforeEach(async () => {
        try {
          await productSeeder(dbConnection)
          await adminSeeder(dbConnection)
          const product = await simpleProductFactory(dbConnection, productData)
          salesChannel = await simpleSalesChannelFactory(dbConnection, {
            name: "test name",
            description: "test description",
            products: [product],
          })
        } catch (err) {
          console.log(err)
          throw err
        }
      })

      afterEach(async () => {
        const db = useDb()
        await db.teardown()
      })

      it("should returns a list of products that belongs to the requested sales channels", async () => {
        const api = useApi()

        const response = await api
          .get(`/admin/products?sales_channel_id[]=${salesChannel.id}`, {
            headers: {
              Authorization: "Bearer test_token",
            },
          })
          .catch((err) => {
            console.log(err)
          })

        expect(response.status).toEqual(200)
        expect(response.data.products.length).toEqual(1)
        expect(response.data.products).toEqual(
          expect.arrayContaining([
            expect.objectContaining({
              id: productData.id,
              title: productData.title,
            }),
          ])
        )
      })
    })
  })
})
