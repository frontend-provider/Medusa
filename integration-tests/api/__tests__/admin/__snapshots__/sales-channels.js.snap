// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sales channels DELETE /admin/sales-channels/:id should delete the requested sales channel 1`] = `
Object {
  "deleted": true,
  "id": Any<String>,
  "object": "sales-channel",
}
`;

exports[`sales channels GET /admin/orders/:id expands sales channel for single 1`] = `
Object {
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "test description",
  "id": Any<String>,
  "is_disabled": false,
  "name": "test name",
  "updated_at": Any<String>,
}
`;

exports[`sales channels GET /admin/orders?expand=sales_channels expands sales channel with parameter 1`] = `
Object {
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "test description",
  "id": Any<String>,
  "is_disabled": false,
  "name": "test name",
  "updated_at": Any<String>,
}
`;

exports[`sales channels GET /admin/sales-channels should list the sales channel 1`] = `
Object {
  "count": 2,
  "limit": 20,
  "offset": 0,
  "sales_channels": ArrayContaining [
    Object {
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "test description",
      "id": Any<String>,
      "is_disabled": false,
      "name": "test name",
      "updated_at": Any<String>,
    },
    Object {
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "test description 2",
      "id": Any<String>,
      "is_disabled": false,
      "name": "test name 2",
      "updated_at": Any<String>,
    },
  ],
}
`;

exports[`sales channels GET /admin/sales-channels should list the sales channel using free text search 1`] = `
Object {
  "count": 1,
  "limit": 20,
  "offset": 0,
  "sales_channels": ArrayContaining [
    Object {
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "test description 2",
      "id": Any<String>,
      "is_disabled": false,
      "name": "test name 2",
      "updated_at": Any<String>,
    },
  ],
}
`;

exports[`sales channels GET /admin/sales-channels should list the sales channel using properties filters 1`] = `
Object {
  "count": 1,
  "limit": 20,
  "offset": 0,
  "sales_channels": ArrayContaining [
    Object {
      "created_at": Any<String>,
      "deleted_at": null,
      "description": "test description",
      "id": Any<String>,
      "is_disabled": false,
      "name": "test name",
      "updated_at": Any<String>,
    },
  ],
}
`;

exports[`sales channels GET /admin/sales-channels/:id should retrieve the requested sales channel 1`] = `
Object {
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "test description",
  "id": Any<String>,
  "is_disabled": false,
  "name": "test name",
  "updated_at": Any<String>,
}
`;

exports[`sales channels POST /admin/sales-channels successfully creates a disabled sales channel 1`] = `
Object {
  "sales_channel": ObjectContaining {
    "is_disabled": true,
    "name": "sales channel name",
  },
}
`;

exports[`sales channels POST /admin/sales-channels successfully creates a sales channel 1`] = `
Object {
  "sales_channel": ObjectContaining {
    "description": "sales channel description",
    "is_disabled": false,
    "name": "sales channel name",
  },
}
`;

exports[`sales channels POST /admin/sales-channels/:id updates sales channel properties 1`] = `
Object {
  "created_at": Any<String>,
  "deleted_at": null,
  "description": "updated description",
  "id": Any<String>,
  "is_disabled": true,
  "name": "updated name",
  "updated_at": Any<String>,
}
`;
